src/backends/plonky2/recursion/circuit.rs

```diff
@@ -59,407 +77,396 @@ pub trait InnerCircuit: Clone {
-            .collect();
-        let verifier_datas: Vec<VerifierCircuitData<F, C, D>> =
-            [verifier_datas, dummy_verifier_datas].concat();
-
-        // set the first n selectors to true, and the rest to false
-        for i in 0..n {
-            pw.set_bool_target(self.targets.selectors_targ[i], true)?;
-        }
-        for i in n..self.params.arity {
-            pw.set_bool_target(self.targets.selectors_targ[i], false)?;
-        }
-
         // set the InnerCircuit related values
-        self.targets
+        self.target
             .innercircuit_targ
-            .set_targets(pw, &innercircuit_input)?;
+            .set_targets(pw, innercircuit_input)?;
 
         #[allow(clippy::needless_range_loop)]
         for i in 0..self.params.arity {
-            pw.set_verifier_data_target(
-                &self.targets.verifier_datas_targ[i],
-                &verifier_datas[i].verifier_only,
-            )?;
-
-            // put together the public inputs with the verifier_data used to
-            // verify the current proof
-            let proof_i_public_inputs = Self::prepare_public_inputs(
-                prev_verifier_datas_hashes[i],
-                recursive_proofs_inp[i].clone(),
-            );
-
-            pw.set_proof_with_pis_target(
-                &self.targets.proofs_targ[i],
-                &ProofWithPublicInputs {
-                    proof: recursive_proofs[i].clone(),
-                    public_inputs: proof_i_public_inputs.clone(),
-                },
-            )?;
+            pw.set_verifier_data_target(&self.target.verifier_datas_targ[i], &verifier_datas[i])?;
+            pw.set_proof_with_pis_target(&self.target.proofs_targ[i], &recursive_proofs[i])?;
         }
 
-        // vds_hash is returned since it will be used as public input to verify
-        // the proof of the current instance of the circuit
-        let vds_hash = hash_verifier_datas(
-            self.params.arity,
-            prev_verifier_datas_hashes.clone(),
-            verifier_datas.clone(),
-        );
-        pw.set_hash_target(self.targets.vds_hash, vds_hash)?;
-
-        Ok(vds_hash)
+        Ok(())
     }
 
     /// returns the full-recursive CircuitData
-    pub fn circuit_data(arity: usize, inner_params: &I::Params) -> Result<CircuitData<F, C, D>> {
-        let data: CircuitData<F, C, D> = common_data_for_recursion::<I>(arity, inner_params)?;
-        let common_data = data.common.clone();
-        let verifier_data = data.verifier_data();
-        let dummy_proof_pi = Self::dummy_proof(data)?;
-        let params = RecursiveParams {
-            arity,
-            common_data,
-            dummy_proof_pi,
-            dummy_verifier_data: verifier_data,
-        };
+    pub fn circuit_data(
+        arity: usize,
+        num_public_inputs: usize,
+        inner_params: &I::Params,
+    ) -> Result<(RecursiveCircuitTarget<I>, CircuitData<F, C, D>)> {
```

Comment:
I was using this method and got confused, since it's named `circuit_data`, but returns both the `RecursiveCircuitTarget` and the `CircuitData`. Maybe it can be reflect that? (also the comment describing it)